### PR TITLE
Glob collections implementation

### PR DESCRIPTION
**Pull request by PapaCharlie**
Implement efficient URI updates by leveraging glob collections. Instead of receiving the entire URI map from the observer every time, this allows receiving only the added/removed URIs in teh cluster. Each cluster is represented as a glob collection named:
```
xdstp:///indis.D2URI/{clusterName}/*
```
And the URI resource names (AKA resource URN) look like this:
```
xdstp:///indis.D2URI/{clusterName}/{query-escaped URI}
```

For testing, I reproduced the same testing steps listed in https://github.com/linkedin/rest.li/pull/975. Here is the attached log for d2-proxy: https://paste.corp.linkedin.com/show/68966295/

The requests are routed correctly, and both symlink and non-symlink clusters seem to be working just fine!

Here are the results of the curli requests: https://paste.corp.linkedin.com/show/68966305/

### CODE REVIEW

File: d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java [hunk 1]
```diff
@@ -533,67 +541,181 @@ public void onReconnect()
...
+          for (Map.Entry<String, XdsD2.D2URI> entry : update.getUris().entrySet())
+          {
+            XdsD2.D2URI d2URI = entry.getValue();
+            UriProperties uriProperties = _uriPropertiesJsonSerializer.fromProto(d2URI);
+            if (uriProperties.getVersion() < 0)
+            {
+              LOG.warn("xDS data for {} has invalid version: {}", entry.getKey(), uriProperties.getVersion());
+            }
+            newData.put(entry.getKey(), uriProperties);
+          }
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java:588]: these has been changed in 29.51.2. Please rebase the latest code.

Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java:588]: This is on top of https://github.com/linkedin/rest.li/pull/975, so I will wait for @brycezhongqing to rebase

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 2]
```diff
@@ -313,36 +364,31 @@ private void handleResourceRemoval(List<String> removedResources, ResourceType t
...
-      case D2_URI_MAP:
-        return _d2URIMapSubscribers;
-      case UNKNOWN:
-      default:
-        throw new AssertionError("Unknown resource type");
-    }
+    return _resourceSubscribers.get(type);
   }
 
   private static final class ResourceSubscriber
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java]: I've rewritten this in the way you suggested, let me know what you think!

File: d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+public class GlobCollectionUtils
+{
+  private static final String D2_URI_NODE_GLOB_COLLECTION_PREFIX = "xdstp:///indis.D2URI/";
+  private static final String GLOB_COLLECTION_SUFFIX = "/*";
+
+  private GlobCollectionUtils()
+  {
+  }
+
+  public static String globCollectionUrlFromResourceUrn(String resourceUrn)
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java:15]: Done!

**Comment by bohhyang** [comment]
For manual testing to avoid regression, could you do the same testing as [this PR](https://github.com/linkedin/rest.li/pull/975) (which verifies SD data fetches by warming up many services including symlink)? you could run it with d2-proxy then send OPTIONs requests to non-symlink d2 endpoints, like toki, and symlink endpoints, like 
`curli -v "d2://entitlementsMaster" --d2-proxy-url "http://localhost:21360/d2/" -X OPTIONS --force-insecure-d2`

File: d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java [hunk 4]
```diff
@@ -626,6 +644,41 @@ private void emitSDStatusUpdateReceiptEvents(Map<String, XdsAndD2Uris> updates,
...
+    }
+
+    @Override
+    public void onReconnect()
+    {
+      _delegate.onReconnect();
+    }
+
+    @Override
+    public void onChanged(XdsClient.D2URICollectionUpdate update)
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java:670]: when the update is empty (new D2URICollectionUpdate()), this method should call `_delegate.onChanged(new XdsClient.D2URIMapUpdate(null));`. Otherwise, the empty map will be published and wipe client side existing data. 

Please add unit tests for this method to cover both normal data and empty data scenarios.

**Comment by bohhyang** [comment]
please also create the container PR to configure the glob collection subscription (could add subscribeToUriGlobCollection to [forceD2ClientConfig](https://jarvis.corp.linkedin.com/codesearch/result/?name=si.src&path=global-config%2Fshared%2Fgroups%2Fei&reponame=linkedin-managed%2Fglobal-config#176)).

Ideally this should be tested against observers that send glob collection responses after [observer PR](https://github.com/linkedin-multiproduct/indis-registry-observer/pull/160) is merged and deployed to EI.

File: d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java [hunk 5]
```diff
@@ -626,6 +644,41 @@ private void emitSDStatusUpdateReceiptEvents(Map<String, XdsAndD2Uris> updates,
...
+    {
+      Map<String, XdsD2.D2URI> copy = new HashMap<>(Maps.transformValues(_delegate._currentData, uri -> uri._xdsUri));
+      copy.putAll(update.getUris());
+      for (String removedUri : update.getRemovedUris())
+      {
+        copy.remove(removedUri);
+      }
+      _delegate.onChanged(new XdsClient.D2URIMapUpdate(copy));
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java:678]: This was pretty much originally your idea!

File: d2/src/main/java/com/linkedin/d2/xds/XdsClient.java [hunk 6]
```diff
@@ -59,19 +63,22 @@ interface D2URIMapResourceWatcher extends ResourceWatcher
...
+  }
+
   interface ResourceUpdate
   {
   }
 
 
   static final class NodeUpdate implements ResourceUpdate
   {
-    String _version;
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClient.java:69]: I wanted to keep it consistent in case we needed it. In the end, we don't really use it and in the case of glob collections, it actually doesn't make much sense since there is no version for the update, only the versions for each URI

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 7]
```diff
@@ -260,22 +269,64 @@ void handleD2URIMapResponse(DiscoveryResponseData data)
...
+        _log.warn("Failed to unpack D2URI glob collection response", e);
+        errors.add("Failed to unpack D2URI glob collection response");
+      }
+    }
+    for (String removedResource : data.getRemovedResources())
+    {
+      getCollectionUpdate(updates, removedResource)
+          .addRemovedUri(removedResource);
+    }
+    sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:313]: we should log about empty D2URICollectionUpdate (similar to line 237 and 270).

```suggestion
    for (Map.Entry<String, D2URICollectionUpdate > entry : updates.entrySet())
    {
       if (isEmptyCollectionUpdate(entry.value()))
       {
          _log.warn("Received D2URICollectionUpdate with no data for resource: {}", entry.key());
       }
    }
    sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
```

isEmptyCollectionUpdate could be a helper function (to be used at line 527 as well).

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 8]
```diff
@@ -260,22 +269,64 @@ void handleD2URIMapResponse(DiscoveryResponseData data)
...
+  {
+    Map<String, D2URICollectionUpdate> updates = new HashMap<>();
+    List<String> errors = new ArrayList<>();
+
+    for (Resource resource : data.getResourcesList())
+    {
+      try
+      {
+        XdsD2.D2URI uri = resource.getResource().unpack(XdsD2.D2URI.class);
+        getCollectionUpdate(updates, resource.getName())
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:299]: instead of passing in the new map "updates", this should be getting the existing update in current subscriber of this glob collection, so that the diffs can be accumulated to the same map to build up the full set of uris of a cluster.

File: d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java [hunk 9]
```diff
@@ -1,41 +1,77 @@
 package com.linkedin.d2.xds;
 
+import javax.annotation.Nullable;
+
 public class GlobCollectionUtils
 {
-  private static final String D2_URI_NODE_GLOB_COLLECTION_PREFIX = "xdstp:///indis.D2URI/";
+  private static final String D2_URIS_PREFIX = "/d2/uris/";
+  private static final String D2_URI_NODE_GLOB_COLLECTION_PREFIX = "xdstp:///indis.D2URI/d2/uris/";
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java]: where does this come from? Isn't d2 uri urn of the format   xdstp:///indis.D2URI/{CLUSTER}/{URL_ESCAPED_ANNOUNCER}?

File: d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java [hunk 10]
```diff
@@ -1,41 +1,77 @@
...
 
-  /**
-   * Extracts the glob collection URL (e.g. xdstp:///indis.D2URI/FooCluster/*) from the given resource URN.
-   */
-  public static String globCollectionUrlFromResourceUrn(String resourceUrn)
+  public static class D2UriIdentifier
   {
-    return resourceUrn.substring(0, resourceUrn.lastIndexOf('/')) + GLOB_COLLECTION_SUFFIX;
-  }
+    private final String _clusterPath;
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java:17]: please add some comment for the expected format of this.

File: d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java [hunk 11]
```diff
@@ -1,41 +1,77 @@
...
+      {
+        return null;
+      }
+      int lastIndex = resourceName.lastIndexOf('/');
+      if (lastIndex == 1)
+      {
+        return null;
+      }
+      return new D2UriIdentifier(
+          D2_URIS_PREFIX + resourceName.substring(0, lastIndex) + GLOB_COLLECTION_SUFFIX,
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java]: this will become "/d2/uris/xdstp:///indis.D2URI/d2/uris/{CLUSTER}/*" then?

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 12]
```diff
@@ -87,35 +87,57 @@ public XdsClientImpl(Node node, ManagedChannel managedChannel, ScheduledExecutor
...
-        subscriber = new ResourceSubscriber(type, resourceName);
+        subscriber = new ResourceSubscriber(watcher.getType(), resourceName);
         resourceSubscriberMap.put(resourceName, subscriber);
+        ResourceType type;
+        String adjustedResourceName;
+        if (watcher.getType() == ResourceType.D2_URI_MAP && _useUriGlobCollections)
+        {
+          type = ResourceType.D2_URI;
+          adjustedResourceName = GlobCollectionUtils.globCollectionUrlForClusterResource(resourceName);
+          _log.info("Glob collection support enabled, subscribing to {} resource: {}", type, adjustedResourceName);
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:125]: I just moved it to the constructor, that way it's logged early.

File: d2/src/main/java/com/linkedin/d2/xds/XdsClient.java [hunk 13]
```diff
@@ -32,47 +30,75 @@
 
 public abstract class XdsClient
 {
-  interface ResourceWatcher
+  public static abstract class ResourceWatcher
   {
+    private final ResourceType _type;
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClient.java:35]: Thanks! Yeah I'm much happier with this, now there's only one `switch` in the code, and all the logic is captured in type-specific implementations

File: d2/src/main/java/com/linkedin/d2/xds/XdsClient.java [hunk 14]
```diff
@@ -32,47 +30,75 @@
...
+      super(ResourceType.NODE);
+    }
+
+    public abstract void onChanged(NodeUpdate update);
 
+    @Override
+    final void onChanged(ResourceUpdate update)
+    {
+      onChanged((NodeUpdate) update);
+    }
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClient.java:78]: Just improves type safety :) Fewer suspicious casts etc

File: d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java (hunk 11 above)
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java]: This has been fixed actually, I think you were looking at an old version

File: d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java (hunk 9 above)
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java]: Yeah this is an old version, I was messing around with stuff. This is gone

File: d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java (hunk 10 above)
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/GlobCollectionUtils.java:17]: Done

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 15]
```diff
@@ -260,22 +312,105 @@ void handleD2URIMapResponse(DiscoveryResponseData data)
...
+        catch (InvalidProtocolBufferException e)
+        {
+          _log.warn("Failed to unpack D2URI glob collection response", e);
+          errors.add("Failed to unpack D2URI glob collection response");
+        }
+      }
+    });
+    sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
+
+    handleResourceUpdate(updates, ResourceType.D2_URI_MAP);
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:409]: if any of the update inside updates has an empty map, we won't be logging about it. Basically the log at line 313 is missing with glob collection enabled.

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 16]
```diff
@@ -260,22 +312,105 @@ void handleD2URIMapResponse(DiscoveryResponseData data)
...
+      {
+        try
+        {
+          XdsD2.D2URI uri = resource.getResource().unpack(XdsD2.D2URI.class);
+          update.getURIMap().put(uriId.getUriName(), uri);
+        }
+        catch (InvalidProtocolBufferException e)
+        {
+          _log.warn("Failed to unpack D2URI glob collection response", e);
+          errors.add("Failed to unpack D2URI glob collection response");
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:403]: Shouldn't mention "glob collection" in the msg, since it's just failing for one D2URI.

File: d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java [hunk 17]
```diff
@@ -57,240 +57,319 @@ public class TestXdsClientImpl
...
+  @Test
+  public void testHandleD2URICollectionUpdateWithEmptyResponse()
   {
     XdsClientImplFixture fixture = new XdsClientImplFixture();
-    fixture._uriMapResourceSubscriber.setData(D2_URI_MAP_UPDATE_WITH_DATA1);
-    fixture.getSpiedXdsClientImpl().handleD2URIMapResponse(DISCOVERY_RESPONSE_URI_MAP_DATA_WITH_REMOVAL);
-    verify(fixture._xdsClientImpl).handleResourceUpdate(eq(new HashMap<>()), eq(D2_URI_MAP));
-    verify(fixture._xdsClientImpl).sendAckOrNack(any(), any(), any());
-    verify(fixture._xdsClientImpl).getResourceSubscriberMap(D2_URI_MAP);
-    verify(fixture._uriMapResourceSubscriber).notifyWatcher(eq(fixture._resourceWatcher), eq(D2_URI_MAP_UPDATE_WITH_DATA1));
```
[the remark this answers is not in the document]
Comment (bohhyang) [reply d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java:289]: nvm. It was moved to lower part.

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java (hunk 15 above)
Comment (bohhyang) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:409]: synced offline. Actually empty maps should be valid for cluster uris maps since the uris under a cluster could have all de-announced legitimately. (as opposed to removing the UriMap resource entirely should be ignored).

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java (hunk 7 above)
Comment (bohhyang) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:407]: synced offline. Actually empty maps should be valid for cluster uris maps since the uris under a cluster could have all de-announced legitimately. (as opposed to removing the UriMap resource entirely should be ignored).

File: d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java [hunk 18]
```diff
@@ -251,23 +252,35 @@ public Object[][] provideBadD2URIMapDataTestCases()
   }
 
   @Test(dataProvider = "badD2URIMapUpdateTestCases")
```
Comment (bohhyang) [inline d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java:254]: the var name at line 82 for the empty data case can be DISCOVERY_RESPONSE_URI_MAP_EMPTY_DATA then.

File: d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java [hunk 19]
```diff
@@ -311,11 +324,12 @@ public void testHandleD2URICollectionResponseWithData()
...
-    actualData = (XdsClient.D2URIMapUpdate) fixture._clusterSubscriber.getData();
+    fixture._xdsClientImpl.handleResponse(createUri2Delete1);
+    actualData = (D2URIMapUpdate) fixture._clusterSubscriber.getData();
     // subscriber data should be updated to D2_URI_MAP_UPDATE_WITH_DATA2
-    verify(fixture._resourceWatcher).onChanged(eq(D2_URI_MAP_UPDATE_WITH_DATA2));
-    Assert.assertEquals(actualData.getURIMap(), D2_URI_MAP_UPDATE_WITH_DATA2.getURIMap());
+    D2URIMapUpdate expectedUpdate = new D2URIMapUpdate(Collections.singletonMap(URI2, URI_BUILDER2.build()));
+    verify(fixture._resourceWatcher).onChanged(eq(expectedUpdate));
+    Assert.assertEquals(actualData.getURIMap(), expectedUpdate.getURIMap());
     fixture.verifyAckSent(2);
```
Comment (bohhyang) [inline d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java:333]: could add another case that if Uri2 is deleted next (the map becomes empty), the watcher will receive an empty map.

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 20]
```diff
@@ -636,6 +700,22 @@ public String toString()
       return "DiscoveryResponseData{" + "_resourceType=" + _resourceType + ", _resources=" + _resources + ", _nonce='"
           + _nonce + '\'' + '}';
     }
+
+    /**
+     * Invokes the given consumer for each resource in this response. If the {@link Resource} is not null, it is being
+     * created/modified and if it is null, it is being removed.
+     */
+    void foreach(BiConsumer<String, Resource> consumer)
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:708]: This was meant to mimic the `forEach` method that's available on `List` and `Map`. `consume` here does somehow imply that a subsequent call to `consumeResource` would yield nothing since they were consumed by the previous call. I changed it to `forEach`, no `foreach` to make it clearer, but I can change it to something else if you want

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java (hunk 16 above)
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:403]: Done

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 21]
```diff
@@ -260,22 +312,105 @@ void handleD2URIMapResponse(DiscoveryResponseData data)
...
+        errors.add(msg);
+        return;
+      }
+
+      ResourceSubscriber subscriber =
+          getResourceSubscriberMap(ResourceType.D2_URI_MAP).get(uriId.getClusterResourceName());
+      if (subscriber == null)
+      {
+        String msg =
+            String.format("Ignoring D2URI resource update for untracked cluster: %s", uriId.getClusterResourceName());
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:358]: Done

File: d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java (hunk 18 above)
Comment (PapaCharlie) [reply d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java:254]: Done

File: d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java (hunk 19 above)
Comment (PapaCharlie) [reply d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java:333]: Done!

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java [hunk 22]
```diff
@@ -101,7 +101,7 @@ public XdsClientImpl(Node node, ManagedChannel managedChannel, ScheduledExecutor
     _node = node;
     _managedChannel = managedChannel;
     _executorService = executorService;
-    _useUriGlobCollections = useUriGlobCollections;
+    _useUriGlobCollections = true;
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:104]: make sure not to check this in:)

File: d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java [hunk 23]
```diff
@@ -331,6 +331,17 @@ public void testHandleD2URICollectionResponseWithData()
     verify(fixture._resourceWatcher).onChanged(eq(expectedUpdate));
     Assert.assertEquals(actualData.getURIMap(), expectedUpdate.getURIMap());
     fixture.verifyAckSent(2);
+
+    // Finally sanity check that the client correctly handles the deletion of the final URI in the collection
+    DiscoveryResponseData deleteUri2 =
+        new DiscoveryResponseData(D2_URI, null, Collections.singletonList(URI_URN2), NONCE, null);
+    fixture._xdsClientImpl.handleResponse(deleteUri2);
+    actualData = (D2URIMapUpdate) fixture._clusterSubscriber.getData();
+    // subscriber data should be updated to D2_URI_MAP_UPDATE_WITH_DATA2
```
Comment (bohhyang) [inline d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java:340]: // subscriber data should be updated to **_empty map_**

File: d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java (hunk 22 above)
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java:104]: lol oops

File: d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java (hunk 23 above)
Comment (PapaCharlie) [reply d2/src/test/java/com/linkedin/d2/xds/TestXdsClientImpl.java:340]: Done

File: d2/src/main/java/com/linkedin/d2/balancer/D2ClientConfig.java [hunk 24]
```diff
@@ -208,7 +209,9 @@ public D2ClientConfig()
                  Long xdsStreamReadyTimeout,
                  ExecutorService dualReadNewLbExecutor,
                  String xdsChannelLoadBalancingPolicy,
-                 Map<String, ?> xdsChannelLoadBalancingPolicyConfig)
+                 Map<String, ?> xdsChannelLoadBalancingPolicyConfig,
+                 boolean useGlobCollectionsForUriSubscriptions
```
[the remark this answers is not in the document]
Comment (PapaCharlie) [reply d2/src/main/java/com/linkedin/d2/balancer/D2ClientConfig.java:213]: Sure

**Review by bohhyang — APPROVED** [review]
